### PR TITLE
focal loss

### PR DESCRIPTION
**Pull request by aashokvardhan**

### CODE REVIEW

**Comment by TortoiseHam** [comment]
can you fix the codacy issues?

File: fastestimator/backend/focal_loss.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        raise ValueError(
+            "Value of gamma should be greater than or equal to zero.")
+
+    if alpha:
+        alpha = cast(alpha, y_pred)
+
+    if gamma:
+        gamma = cast(gamma, y_pred)
+
+    if isinstance(y_true, tf.Tensor):
```
Comment (TortoiseHam) [inline fastestimator/backend/focal_loss.py:99]: I think this needs to be tf.is_tensor() rather than isinstance() in order to work with variables / multi -gpu

File: fastestimator/backend/focal_loss.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    if gamma:
+        modulating_factor = (1.0 - p_t)**gamma
+
+    if reduction == "mean":
+        loss = reduce_mean(alpha_factor * modulating_factor * ce)
+    elif reduction == "sum":
+        loss = reduce_sum(alpha_factor * modulating_factor * ce)
+    else:
+        loss = alpha_factor * modulating_factor * ce
```
Comment (TortoiseHam) [inline fastestimator/backend/focal_loss.py:143]:
```suggestion
    loss = alpha_factor * modulating_factor * ce
    if reduction == "mean":
        loss = reduce_mean(loss)
    elif reduction == "sum":
        loss = reduce_sum(loss)
```

File: fastestimator/backend/focal_loss.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import torch
+
+from fastestimator.backend.cast import cast
+from fastestimator.backend.reduce_mean import reduce_mean
+from fastestimator.backend.reduce_sum import reduce_sum
+
+Tensor = TypeVar('Tensor', tf.Tensor, torch.Tensor)
+
+
+def cast(input_data, y_true):
```
Comment (TortoiseHam) [inline fastestimator/backend/focal_loss.py:27]: probably delete this and use the backend cast

File: fastestimator/op/tensorop/loss/focal_loss.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+from fastestimator.util.traceability_util import traceable
+
+Tensor = TypeVar('Tensor', tf.Tensor, torch.Tensor)
+
+
+@traceable()
+class FocalLoss(LossOp):
+    """
+    Calculate Focal Loss.
+    Original implementation from https://github.com/facebookresearch/fvcore/blob/master/fvcore/nn/focal_loss.py .
```
Comment (TortoiseHam) [inline fastestimator/op/tensorop/loss/focal_loss.py:31]:
```suggestion
    """Calculate Focal Loss.

    Original implementation from https://github.com/facebookresearch/fvcore/blob/master/fvcore/nn/focal_loss.py .
```

File: fastestimator/op/tensorop/loss/focal_loss.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    Loss used in RetinaNet for dense detection: https://arxiv.org/abs/1708.02002.
+
+    Args:
+        inputs: A tuple or list like: [<y_pred>, <y_true>].
+        outputs: String key under which to store the computed loss value.
+        alpha: Weighting factor in range (0,1) to balance
+                positive vs negative examples or -1 for ignore. Default = 0.25
```
Comment (TortoiseHam) [inline fastestimator/op/tensorop/loss/focal_loss.py:38]:
```suggestion
                positive vs negative examples or -1 to ignore. Default = 0.25
```

File: fastestimator/backend/focal_loss.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        ce = ce.view(ce.shape[0], -1)
+    else:
+        raise ValueError("Unsupported tensor type.")
+
+    p_t = (y_true * pred_prob) + ((1 - y_true) * (1 - pred_prob))
+
+    alpha_factor = 1.0
+    modulating_factor = 1.0
+
+    if alpha:
```
Comment (TortoiseHam) [inline fastestimator/backend/focal_loss.py:132]: how does alpha of -1 get ignored (as claimed in the doc string)? if alpha will evaluate to True for all cases except alpha=0

File: fastestimator/op/tensorop/loss/focal_loss.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+from typing import Any, Dict, Iterable, List, Tuple, TypeVar, Union
+
+import tensorflow as tf
+import torch
+
+from fastestimator.backend import focal_loss
```
Comment (TortoiseHam) [inline fastestimator/op/tensorop/loss/focal_loss.py:20]: when we import within non-apphub FE code we should use the full path (fastestimator.backend._focal_loss import focal_loss)

File: fastestimator/backend/_focal_loss.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+import tensorflow as tf
+import torch
+
+from fastestimator.backend import reduce_mean, reduce_sum
+
+Tensor = TypeVar('Tensor', tf.Tensor, torch.Tensor)
+
+
+def cast(input_data, y_pred):
```
Comment (TortoiseHam) [inline fastestimator/backend/_focal_loss.py:25]: should we adjust our backend cast function to also take non-string dtypes and then use it here rather than writing another cast function?

File: fastestimator/backend/_focal_loss.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+from typing import TypeVar
+
+import tensorflow as tf
+import torch
+
+from fastestimator.backend import cast, reduce_mean, reduce_sum
```
Comment (TortoiseHam) [inline fastestimator/backend/_focal_loss.py:20]: since this is part of the fe codebase it should import from the absolute path (from fastestimator.backend._cast import cast), which helps prevent cyclical import issues
